### PR TITLE
disable digest resolution for manifest v2 schema 1 manifests

### PR DESCRIPTION
**Pull request by Yiyuanzzz**
### Summary

ECS Agent will disable the digest resolution for Schema 1 manifest as a workaround for a bug in ECR that ECR will reports an incorrect digest during digest resolution phase for Schema V1 manifests which then causes image pull using the digest to fail. 
### Implementation details

skip digest resolution for manifest v2 schema 1 manifests 
### Testing

Reproduce the issue by setting Up Pull-Through Cache Rule
```
aws ecr create-pull-through-cache-rule \
     --ecr-repository-prefix quay \
     --upstream-registry-url quay.io \
     --region us-west-2
docker pull  130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4
```
Manually ran a task with schema 1 image 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4, we can see error " image verification failed for digest "

```
level=debug time=2024-07-22T05:06:56Z msg="Setting image digest on container" taskARN="arn:aws:ecs:us-west-2:130669029530:task/tssbug/ca80402ec1ff44e398c7a88bef4bef7f" containerName="container" imageDigest="sha256:d1f9e51bf7f3f7940faf11553d08f8815d2f0f999f6eafd6a6b8f765adba5895"
level=error time=2024-07-22T05:07:07Z msg="DockerGoClient: failed to pull image 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4@sha256:d1f9e51bf7f3f7940faf11553d08f8815d2f0f999f6eafd6a6b8f765adba5895: [CannotPullContainerError] image verification failed for digest sha256:d1f9e51bf7f3f7940faf11553d08f8815d2f0f999f6eafd6a6b8f765adba5895" module=docker_client.go
```
Manually ran a task with schema 1 image 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4 with  test agent(changes of this pr), we can see digest resolution is skipped and pulling image complete.

```
level=info time=2024-07-22T05:14:24Z msg="skipping digest resolution for manifest v2 schema 1," taskARN="arn:aws:ecs:us-west-2:130669029530:task/default/eb9cf2168d9c406388966b6e2584dae5" containerName="container" ImageMediatype="application/vnd.docker.distribution.manifest.v1+prettyjws" image="130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4"
l

level=debug time=2024-07-22T05:14:25Z msg="DockerGoClient: pulling image 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4, status Status: Downloaded newer image for 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4" module=docker_client.go
level=debug time=2024-07-22T05:14:25Z msg="DockerGoClient: pulling image complete: 130669029530.dkr.ecr.us-west-2.amazonaws.com/quay/coreos/etcd:v2.0.4" module=docker_client.go
```

New tests cover the changes: 

### Description for the changelog

Bugfix: Disable Digest Resolution for Manifest V2 Schema 1

**Does this PR include breaking model changes? If so, Have you added transformation functions?**
  

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_task_engine_test.go [hunk 1]
```diff
@@ -4229,7 +4246,12 @@ func TestPullContainerManifest(t *testing.T) {
 				versioned.EXPECT().
 					PullImageManifest(gomock.Any(), "myimage", nil).
 					Return(
-						registry.DistributionInspect{Descriptor: ocispec.Descriptor{Digest: testDigest}},
+						registry.DistributionInspect{
+							Descriptor: ocispec.Descriptor{
+								MediaType: "application/vnd.docker.distribution.manifest.v2+json",
```
Comment (amogh09) [inline agent/engine/docker_task_engine_test.go:4251]: nit: maybe we should define a constant for v2 schema and use it here.

File: ecs-agent/logger/field/constants.go [hunk 2]
```diff
@@ -49,6 +49,7 @@ const (
 	ImageLastUsedAt         = "imageLastUsedAt"
 	ImagePullSucceeded      = "imagePullSucceeded"
 	ImageDigest             = "imageDigest"
+	ImageMediatype          = "imageMediatype"
```
Comment (amogh09) [inline ecs-agent/logger/field/constants.go:52]: nit: `Mediatype` should be `MediaType`

Comment (Yiyuanzzz) [reply ecs-agent/logger/field/constants.go]: thanks! updated

File: agent/engine/docker_task_engine_test.go (hunk 1 above)
Comment (Yiyuanzzz) [reply agent/engine/docker_task_engine_test.go]: updated

File: agent/engine/docker_task_engine.go [hunk 3]
```diff
@@ -1352,6 +1353,22 @@ func (engine *DockerTaskEngine) pullContainerManifest(
...
 			return dockerapi.DockerContainerMetadata{Error: manifestPullErr}
 		}
+		imageManifestMediatype := distInspect.Descriptor.MediaType
+		logger.Info("Fetched image manifest Mediatype for container from registry", logger.Fields{
+			field.TaskARN:        task.Arn,
+			field.ContainerName:  container.Name,
+			field.ImageMediatype: imageManifestMediatype,
+			field.Image:          container.Image,
+		})
+		if strings.Contains(imageManifestMediatype, "application/vnd.docker.distribution.manifest.v1") {
```
[the remark this answers is not in the document]
Comment (Yiyuanzzz) [reply agent/engine/docker_task_engine.go]: thanks! updated

File: agent/engine/docker_task_engine.go [hunk 4]
```diff
@@ -1352,6 +1358,22 @@ func (engine *DockerTaskEngine) pullContainerManifest(
...
 			return dockerapi.DockerContainerMetadata{Error: manifestPullErr}
 		}
+		imageManifestMediaType := distInspect.Descriptor.MediaType
+		logger.Info("Fetched image manifest MediaType for container from registry", logger.Fields{
+			field.TaskARN:        task.Arn,
+			field.ContainerName:  container.Name,
+			field.ImageMediaType: imageManifestMediaType,
+			field.Image:          container.Image,
+		})
+		if strings.Contains(imageManifestMediaType, mediaTypeManifestV1) || strings.Contains(imageManifestMediaType, mediaTypeSignedManifestV1) {
```
Comment (amogh09) [inline agent/engine/docker_task_engine.go:1368]: Why do we want to do `strings.Contains` instead of just `==`?

**Review by amogh09 — APPROVED** [review]

**Review by tinnywang — APPROVED** [review]

**Pull request merged**